dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,337 @@
+package org.checkerframework.dataflow.cfg;
+
+import java.util.*;
+import javax.lang.model.type.TypeMirror;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.dataflow.analysis.*;
+import org.checkerframework.dataflow.cfg.block.*;
+import org.checkerframework.dataflow.cfg.node.Node;
+
+/**
+ * An abstract class of control flow graph visualizer. To achieve this abstract class, you need to
+ * implement some of the methods in {@link CFGVisualizer}.
+ *
+ * <p>Examples: {@link DOTCFGVisualizer} and {@link StringCFGVisualizer}.
+ */
+public abstract class AbstractCFGVisualizer<
+                A extends AbstractValue<A>, S extends Store<S>, T extends TransferFunction<A, S>>
+        implements CFGVisualizer<A, S, T> {
+
+    protected boolean verbose;
+
+    @Override
+    public void init(Map<String, Object> args) {
+        {
+            Object verb = args.get("verbose");
+            this.verbose =
+                    verb != null
+                            && (verb instanceof String
+                                    ? Boolean.getBoolean((String) verb)
+                                    : (boolean) verb);
+        }
+    }
+
+    protected String generateGraphHelper(
+            ControlFlowGraph cfg,
+            Block entry,
+            @Nullable Analysis<A, S, T> analysis,
+            String thenLabelContent,
+            String elseLabelContent) {
+        Set<Block> visited = new HashSet<>();
+
+        StringBuilder sbDigraph = new StringBuilder();
+        sbDigraph.append("digraph {\n");
+
+        Block cur = entry;
+        Queue<Block> workList = new ArrayDeque<>();
+        visited.add(entry);
+        // traverse control flow graph and define all arrows
+        while (cur != null) {
+            if (cur.getType() == Block.BlockType.CONDITIONAL_BLOCK) {
```

Comment:
Would it make the code easier to read if you extract the whole body of the `while`, except for the `cur = workList.poll();` into a separate helper? That would make the overall structure of this method clearer and the separate method isn't as nested.

Reply:
When I am trying to achieve this, I found that I needed to maintain `visited`, `workList` and `sbDigraph` at the same time. Since we do not want to use any global variables, I think it is a bit troublesome. Please give me some comments.